### ISSUE TITLE
RangeError: Maximum call stack size exceeded, from no-useless-return rule

### ISSUE THREAD

**Issue by AlexTes** [eslint/eslint#7855]
**Tell us about your environment**

* **ESLint Version:** v3.12.2
* **Node Version:** v7.3.0
* **npm Version:** 4.0.5

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
```
{
  "root": true,
  "env": {
    "node": true
  },
  "parserOptions": {
      "ecmaVersion": 6
  },
  "rules": {
    "no-useless-return": "error"
  }
}
```

**What did you do? Please include the actual source code causing the issue.**
Trying to test a catch block. At the same time making sure I'm writing code that's linted (naturally).
```js
try {
  throw new Error('intentional throw');
  while (false) {}
} catch (err) {
  // do nothing
}
```

**What did you expect to happen?**
For eslint to lint my code successfully. Perhaps reporting an unreachable code error if the rule is turned on. Perhaps eslint feels like positively surprising me with a no-throw-in-try-catch I didn't know about.

**What actually happened? Please include the actual, raw output from ESLint.**
Eslint encounters an unexpected error. Output with debug turned on.
```
$ eslint --debug index.js
  eslint:cli Running on files +0ms
  eslint:glob-util Creating list of files to process. +39ms
  eslint:ignored-paths Looking for ignore file in /Users/alexander/code/no-useless-return +4ms
  eslint:ignored-paths Could not find ignore file in cwd +1ms
  eslint:cli-engine Processing /Users/alexander/code/no-useless-return/index.js +5ms
  eslint:cli-engine Linting /Users/alexander/code/no-useless-return/index.js +1ms
  eslint:config Constructing config for /Users/alexander/code/no-useless-return/index.js +0ms
  eslint:config Using .eslintrc and package.json files +0ms
  eslint:config Loading /Users/alexander/code/no-useless-return/.eslintrc +2ms
  eslint:config-file Loading config file: /Users/alexander/code/no-useless-return/.eslintrc +1ms
  eslint:config Using /Users/alexander/code/no-useless-return/.eslintrc +59ms
  eslint:config Merging command line environment settings +1ms
  eslint:config-ops Apply environment settings to config +0ms
  eslint:config-ops Creating config for environment node +0ms
Maximum call stack size exceeded
RangeError: Maximum call stack size exceeded
    at isReturned (/Users/alexander/.nvm/versions/node/v7.3.0/lib/node_modules/eslint/lib/rules/no-useless-return.js:119:28)
    at Array.filter (native)
    at getUselessReturns (/Users/alexander/.nvm/versions/node/v7.3.0/lib/node_modules/eslint/lib/rules/no-useless-return.js:148:49)
    at getUselessReturns (/Users/alexander/.nvm/versions/node/v7.3.0/lib/node_modules/eslint/lib/rules/no-useless-return.js:146:21)
    at getUselessReturns (/Users/alexander/.nvm/versions/node/v7.3.0/lib/node_modules/eslint/lib/rules/no-useless-return.js:146:21)
    at getUselessReturns (/Users/alexander/.nvm/versions/node/v7.3.0/lib/node_modules/eslint/lib/rules/no-useless-return.js:146:21)
    at getUselessReturns (/Users/alexander/.nvm/versions/node/v7.3.0/lib/node_modules/eslint/lib/rules/no-useless-return.js:146:21)
    at getUselessReturns (/Users/alexander/.nvm/versions/node/v7.3.0/lib/node_modules/eslint/lib/rules/no-useless-return.js:146:21)
    at getUselessReturns (/Users/alexander/.nvm/versions/node/v7.3.0/lib/node_modules/eslint/lib/rules/no-useless-return.js:146:21)
    at getUselessReturns (/Users/alexander/.nvm/versions/node/v7.3.0/lib/node_modules/eslint/lib/rules/no-useless-return.js:146:21)
```

To be clear, I'm not sure this qualifies as a bug. One could argue you should never throw in a try / catch block. Moving the `throw` statement into an if-statement, or a separate function resolves the error. Since that will be the case normally this is not much of a bug. The argument in favor I can see is that this piece of code makes sense when trying to check - without writing a proper test - the catch block.

I tinkered with the code quite a bit in an attempt to reduce the amount of time you lovely people will have to spend on it. The `try...catch` is essential. The `throw` statement being made directly in the `try` block appears essential (tried a few variants of moving it to a different block / closure, all mitigate the issue). The `while` loop is essential. Removing the `catch` block gives an error about the `try` block missing an `catch` or `finally` block.

Hope you feel this is helpful!

**Comment by platinumazure** [comment]
Where possible, we never want rules to crash-- if ESLint has an opinion about some bit of code, that should be expressed via a lint report in an ESLint run that eventually halts. So this is definitely a bug!

**Comment by platinumazure** [comment]
Definitely was able to soft-crash the demo page with this example:

```js
/* eslint no-useless-return:error, no-unreachable:off, no-empty:off, no-constant-condition:off */
try {
  throw new Error('intentional throw');
  while (false) {}
} catch (err) {
  // do nothing
}
```

After it tries linting that, it won't re-lint the code even if I make changes in the demo.

**Comment by AlexTes** [comment]
Glad to hear platinumazure 👍 
Also hi not-an-aardvark, nice to run into you again ^_^. Going to have a look at your PR and see if maybe next time I should try myself :]
